lib/matplotlib/image.py

```diff
@@ -1514,18 +1514,36 @@ def imsave(fname, arr, vmin=None, vmax=None, cmap=None, format=None,
             origin = mpl.rcParams["image.origin"]
         if origin == "lower":
             arr = arr[::-1]
-        rgba = sm.to_rgba(arr, bytes=True)
+        if (isinstance(arr, memoryview) and arr.format == "B"
+                and arr.ndim == 3 and arr.shape[-1] == 4):
+            # Such an ``arr`` would also be handled fine by sm.to_rgba (after
+            # casting with asarray), but it is useful to special-case it
+            # because that's what backend_agg passes, and can be in fact used
+            # as is, saving a few operations.
+            rgba = arr
+        else:
+            rgba = sm.to_rgba(arr, bytes=True)
         if pil_kwargs is None:
             pil_kwargs = {}
         pil_shape = (rgba.shape[1], rgba.shape[0])
         image = PIL.Image.frombuffer(
             "RGBA", pil_shape, rgba, "raw", "RGBA", 0, 1)
-        if format == "png" and metadata is not None:
-            # cf. backend_agg's print_png.
-            pnginfo = PIL.PngImagePlugin.PngInfo()
-            for k, v in metadata.items():
-                pnginfo.add_text(k, v)
-            pil_kwargs["pnginfo"] = pnginfo
+        if format == "png":
+            # Only use the metadata kwarg if pnginfo is not set, because the
+            # semantics of duplicate keys in pnginfo is unclear.
+            if "pnginfo" in pil_kwargs:
+                if metadata:
+                    cbook._warn_external("'metadata' is overridden by the "
+                                         "'pnginfo' entry in 'pil_kwargs'.")
+            else:
+                metadata = {
+                    "Software": (f"matplotlib version{mpl.__version__}, "
+                                 f"http://matplotlib.org/"),
+                    **(metadata if metadata is not None else {}),
```

Comment:
This `is not None` means you can't override the `Software` key.

Reply:
You can?  Just pass whatever you want as Software key in metadata?

Reply:
By override, I meant, remove.

Reply:
1) Isn't that the same as before?
2) You can always set it to an empty string, is that functionally different?

Reply:
Ah, you're right; I guess this is something only the PDF backends do?
https://github.com/matplotlib/matplotlib/blob/793c6b05381231371267b44b107726f3878e14f2/lib/matplotlib/backends/backend_pdf.py#L522-L523
Since this was passed down, it looks like previously it would send `None` to `add_text`, and looking at Pillow, that would just crash, since it expects `str`/`bytes`.

Reply:
I can handle None as "remove the key", if you want?

Reply:
Yes, I think so. While PDF has more default keys, so it makes more sense there, I think we should have consistent behaviour and do it for PNG as well.

Reply:
Actually I guess the patch is just
```patch
diff --git i/lib/matplotlib/image.py w/lib/matplotlib/image.py
index 51f8a3982..803baa673 100644
--- i/lib/matplotlib/image.py
+++ w/lib/matplotlib/image.py
@@ -1543,7 +1543,8 @@ def imsave(fname, arr, vmin=None, vmax=None, cmap=None, format=None,
                 }
                 pil_kwargs["pnginfo"] = pnginfo = PIL.PngImagePlugin.PngInfo()
                 for k, v in metadata.items():
-                    pnginfo.add_text(k, v)
+                    if v is not None:
+                        pnginfo.add_text(k, v)
         if format in ["jpg", "jpeg"]:
             format = "jpeg"  # Pillow doesn't recognize "jpg".
             facecolor = mpl.rcParams["savefig.facecolor"]
```
but the behavior would be slightly annoying to document, do you want to give it a try?

Reply:
For PDF, it just says: "The standard keys are ... Values have been predefined for ... They can be removed by setting them to `None`." So I guess something simple like:
```patch
diff --git a/lib/matplotlib/backends/backend_agg.py b/lib/matplotlib/backends/backend_agg.py
index 67e28efd2..0fd6d52a7 100644
--- a/lib/matplotlib/backends/backend_agg.py
+++ b/lib/matplotlib/backends/backend_agg.py
@@ -491,8 +491,8 @@ class FigureCanvasAgg(FigureCanvasBase):
 
             Other keywords may be invented for other purposes.
 
-            If 'Software' is not given, an autogenerated value for matplotlib
-            will be used.
+            If 'Software' is not given, an autogenerated value for Matplotlib
+            will be used. This can be removed by setting it to *None*.
 
             For more details see the `PNG specification`_.
```

Reply:
done